storage/innobase/buf/buf0flu.cc

```diff
@@ -1903,6 +1903,22 @@ ATTRIBUTE_COLD static void buf_flush_sync_for_checkpoint(lsn_t lsn)
   }
 }
 
+/** Check if the adpative flushing threshold is recommended based on
+redo log capacity filled threshold.
+@param oldest_lsn     buf_pool.get_oldest_modification(0)
+@return true if adaptive flushing is recommended. */
+static
+bool
+af_needed_based_on_redo_activity(lsn_t oldest_lsn)
+{
+  lsn_t age= (log_sys.get_lsn() - oldest_lsn);
+  lsn_t af_lwm= static_cast<lsn_t>(srv_adaptive_flushing_lwm *
+   static_cast<double>(log_sys.log_capacity) / 100);
+
+  /* if age > af_lwm adaptive flushing is recommended */
+  return (age > af_lwm);
+}
```

Comment:
The comment on `oldest_lsn` suggests that we could pass `oldest_lsn==0`. My suggestion for this function would be as follows:
```c++
/** Check if adaptive flushing is recommended
@param oldest_lsn  buf_pool.get_oldest_modified()->oldest_modification()
@return whether adaptive flushing is recommended */
static bool af_needed(lsn_t oldest_lsn)
{
  return srv_adaptive_flushing &&
    (log_sys.get_lsn() - oldest_lsn) >
    static_cast<lsn_t>(log_sys.log_capacity * srv_adaptive_flushing_lwm / 100);
}
```
With the default settings, this should make no difference, because `srv_adaptive_flushing` (`innodb_adaptive_flushing`) is enabled by default.

Reply:
Sorry, my request regarding `srv_adaptive_flushing &&` must be discarded. I was not aware of this constraint that MariaDB 10.2 through 10.4 have inherited [from MySQL 5.7](https://dev.mysql.com/doc/refman/5.7/en/innodb-buffer-pool-flushing.html):
>The `innodb_adaptive_flushing_lwm` variable defines a low water mark for redo log capacity. When that threshold is crossed, adaptive flushing is enabled, even if the `innodb_adaptive_flushing` variable is disabled.

Perhaps @iangilfillan could ensure that this mentioned in the MariaDB documentation, just like the special meaning of `innodb_max_dirty_pages_pct_lwm=0` (MDEV-24537).

We should not change such logic within a GA release. We might revise this decision in a later development series that has not reached GA status yet.

Reply:
As per the original semantics (please refer to mysql documentation) https://dev.mysql.com/doc/refman/8.0/en/innodb-buffer-pool-flushing.html

The innodb_adaptive_flushing_lwm variable defines a low water mark for redo log capacity. When that threshold is crossed, adaptive flushing is enabled, even if the innodb_adaptive_flushing variable is disabled.

So srv_adaptive_flushing doesn't make difference if the threshold is crossed.